pkg/util/goroutinemap/goroutinemap.go

```diff
@@ -45,58 +56,133 @@ type GoRoutineMap interface {
+		exponentialBackOffOnError: exponentialBackOffOnError,
 	}
 }
 
 type goRoutineMap struct {
-	operations map[string]bool
+	operations                map[string]operation
+	exponentialBackOffOnError bool
+	wg                        sync.WaitGroup
 	sync.Mutex
-	wg sync.WaitGroup
 }
 
-func (grm *goRoutineMap) Run(operationName string, operation func() error) error {
+type operation struct {
+	operationPending    bool
+	lastError           error
+	lastErrorTime       time.Time
+	durationBeforeRetry time.Duration
+}
+
+func (grm *goRoutineMap) Run(operationName string, operationFunc func() error) error {
 	grm.Lock()
 	defer grm.Unlock()
-	if grm.operations[operationName] {
+	existingOp, exists := grm.operations[operationName]
+	if exists {
 		// Operation with name exists
-		return newAlreadyExistsError(operationName)
+		if existingOp.operationPending {
+			return newAlreadyExistsError(operationName)
+		}
+
+		if time.Since(existingOp.lastErrorTime) <= existingOp.durationBeforeRetry {
+			return newExponentialBackoffError(operationName, existingOp)
+		}
 	}
 
-	grm.operations[operationName] = true
+	grm.operations[operationName] = operation{
+		operationPending:    true,
+		lastError:           existingOp.lastError,
+		lastErrorTime:       existingOp.lastErrorTime,
+		durationBeforeRetry: existingOp.durationBeforeRetry,
+	}
 	grm.wg.Add(1)
-	go func() {
-		defer grm.operationComplete(operationName)
-		defer runtime.HandleCrash()
-		operation()
+	go func() (err error) {
+		// Handle unhandled panics (very unlikely)
+		defer k8sRuntime.HandleCrash()
+		// Handle completion of and error, if any, from operationFunc()
+		defer grm.operationComplete(operationName, &err)
+		// Handle panic, if any, from operationFunc()
+		defer recoverFromPanic(operationName, &err)
+		return operationFunc()
 	}()
 
 	return nil
 }
 
-func (grm *goRoutineMap) operationComplete(operationName string) {
+func (grm *goRoutineMap) operationComplete(operationName string, err *error) {
 	defer grm.wg.Done()
 	grm.Lock()
 	defer grm.Unlock()
-	delete(grm.operations, operationName)
+
+	if *err == nil || !grm.exponentialBackOffOnError {
+		// Operation completed without error, or exponentialBackOffOnError disabled
+		delete(grm.operations, operationName)
+		if *err != nil {
+			// Log error
+			glog.Errorf("operation for %q failed with: %v",
+				operationName,
+				*err)
+		}
+	} else {
+		// Operation completed with error
+		existingOp := grm.operations[operationName]
+		if existingOp.durationBeforeRetry == 0 {
+			existingOp.durationBeforeRetry = initialDurationBeforeRetry
+		} else {
+			existingOp.durationBeforeRetry = 2 * existingOp.durationBeforeRetry
+		}
+		existingOp.lastError = *err
+		existingOp.lastErrorTime = time.Now()
+		existingOp.operationPending = false
+
+		grm.operations[operationName] = existingOp
+		if *err != nil {
+			// Log error
+			glog.Errorf("Operation for %q failed. No retries permitted until %v (%v). error: %v",
```

Comment:
Operation for %q failed. No retries permitted until %v (durationBeforeRetry %v). error: %v",